### ISSUE TITLE
Replace "httplib/m2crypto" with "requests

### ISSUE THREAD

**Issue by jealous** [pywbem/pywbem#280]
We are having trouble install pywbem on a fresh installed ubuntu.
The root cause is pywbem has dependency to m2crypto which turns out to have a bunch of dependencies that cannot be handled by "pip".
Could we just replace the httplib with the more popular "requests" lib?  As far as I could see, it greatly reduces the complexity of `cim_http.wbem_request` and it removes the dependency of m2crypto which make the installation on multiple platforms much easier than before.

Here are some details of the installation error.

```
building 'M2Crypto.__m2crypto' extension

swigging SWIG/_m2crypto.i to SWIG/_m2crypto_wrap.c

swig -python -D__x86_64__ -I/usr/include/python2.7 -I/usr/include -I/usr/include/x86_64-linux-gnu -includeall -modern -builtin -outdir build/lib.linux-x86_64-2.7/M2Crypto -o SWIG/_m2crypto_wrap.c SWIG/_m2crypto.i

unable to execute 'swig': No such file or directory

error: command 'swig' failed with exit status 1
```

This error could be resolved by `apt-get install swig`
Then I received many warnings from SWIG.  Here is just a part of it.

```
    SWIG/_bio.i:64: Warning 454: Setting a pointer/reference variable may leak memory.
    SWIG/_rand.i:21: Warning 454: Setting a pointer/reference variable may leak memory.
    SWIG/_evp.i:169: Warning 454: Setting a pointer/reference variable may leak memory.
    SWIG/_dh.i:36: Warning 454: Setting a pointer/reference variable may leak memory.
    SWIG/_rsa.i:43: Warning 454: Setting a pointer/reference variable may leak memory.
    SWIG/_dsa.i:31: Warning 454: Setting a pointer/reference variable may leak memory.
    SWIG/_ssl.i:241: Warning 454: Setting a pointer/reference variable may leak memory.
    SWIG/_ssl.i:242: Warning 454: Setting a pointer/reference variable may leak memory.
    SWIG/_x509.i:323: Warning 454: Setting a pointer/reference variable may leak memory.
    SWIG/_pkcs7.i:44: Warning 454: Setting a pointer/reference variable may leak memory.
    SWIG/_pkcs7.i:44: Warning 454: Setting a pointer/reference variable may leak memory.
    SWIG/_util.i:11: Warning 454: Setting a pointer/reference variable may leak memory.
    SWIG/_ec.i:111: Warning 454: Setting a pointer/reference variable may leak memory.
    SWIG/_engine.i:168: Warning 454: Setting a pointer/reference variable may leak memory.
    x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -I/usr/include/python2.7 -I/usr/include -I/usr/include/x86_64-linux-gnu -I/tmp/pip_build_root/M2Crypto/SWIG -c SWIG/_m2crypto_wrap.c -o build/temp.linux-x86_64-2.7/SWIG/_m2crypto_wrap.o -DTHREADING
    SWIG/_m2crypto_wrap.c:7121:2: warning: function declaration isn’t a prototype [-Wstrict-prototypes]
     *x509v3_lhash() {
      ^
    SWIG/_m2crypto_wrap.c: In function ‘x509v3_lhash’:
    SWIG/_m2crypto_wrap.c:7122:5: warning: return from incompatible pointer type [enabled by default]
         return lh_new(NULL, NULL); /* Should probably be lh_CONF_VALUE_new but won't compile. */
         ^
    SWIG/_m2crypto_wrap.c: In function ‘_wrap_new__STACK’:
    SWIG/_m2crypto_wrap.c:8379:1: warning: label ‘fail’ defined but not used [-Wunused-label]
     fail:
     ^
    SWIG/_m2crypto_wrap.c: In function ‘_wrap_sk_new_null’:
    SWIG/_m2crypto_wrap.c:8522:1: warning: label ‘fail’ defined but not used [-Wunused-label]
     fail:
```

And someone else in our team met a openssl header issue later.

**Comment by jealous** [comment]
More information.  Due to the dependency to `swig`, our package which utilize pywbem cannot pass travis CI or openstack CI.
See following build result:

https://travis-ci.org/emc-openstack/storops/jobs/127763475

**Comment by andy-maier** [comment]
Hi Cedric.
Karl and I agree that going to the `requests` package would be a good move, and we plan that for the version after 0.9.0.

For now, I'd like to point you to the already available possibility of letting the PyWBEM installation automatically establish its OS-level package prerequisites such as Swig, or OpenSSL header issues. The automatic install is supported for a number of well-known Linux distros, including Ubuntu. (Just for completeness: For the other distros, it displays the RHEL package names that need to be installed, so the user needs to translate them to the actual Linux distro and install themselves).

The free version of Travis runs Ubuntu 12.04, which does not have a Swig package. In that case,
the automatic support of PyWBEM builds Swig from its sources, and establishes the development prerequisites that are needed for that. Installing M2Crypto is a challenging task, but PyWBEM has pretty much fully automated that by now.

If you don't trust that, let me tell you that the Travis runs of PyWBEM itself uses this approach to get m2Crypto installed, and it works great.

I have created PR #282 which updates the INSTALL.md file to better describe how that is done.
In short, here are the commands I would recommend for your case (they can be added to the Travis control file, e.g. into the `install` section, and `cd` does work there):

      pip download pywbem
      tar -xf pywbem-0.8.3.tar.gz
      cd pywbem-0.8.3
      python setup.py install_os install

You can also try to use wildcards instead of specifying the exact version, but I did not verify whether they will be properly expanded in Travis.

Please let me know whether you can use this approach.

**Comment by jealous** [comment]
Hi Andy,

Thank you very much for the quick and detailed response!

For a temporary solution, we are changing our dependency to a fork of pywbem and our travis build completed successfully.  But in the long run, we will definitely switch back to pywbem when it's moved to requests.  The reason is simple, we don't have the resource to maintain an up-to-date fork and you guys are the experts for this project and has much more tests to cover the whole project.

I will try the travis solution tomorrow and I believe it will work as you described.

But since our package need to support windows as well, I don't think we could easily move back when the dependency to swig and openssl headers still exists.  Because it brought much more trouble to the user than the *nix platforms.

**Comment by andy-maier** [comment]
The remainder of this issue is for pywbem to migrate to the `requests` module.

Changing target milestone to 0.10.0.

**Comment by skolsuper** [comment]
I am not the issue originator, but my company was also using a fork of this project for a long time.  However as of yesterday I have switched back to the official `0.8.3` release, by subclassing the WBEMConnection class as follows:

```python
class AsyncWBEMConnection(pywbem.WBEMConnection):

    async def wbem_request(self, data: str, headers: Dict[str, str], timeout: Number, **kwargs):
        """
        Execute the CIM-XML HTTP request in a non-blocking manner.
        Args:
            data: The encoded (content) XML request
            headers: Any extra headers besides the defaults set below.
            timeout: Timeout for the request in seconds.
            **kwargs: There are numerous extra arguments passed to this function that I just choose to ignore.
                      It doesn't seem to be a problem.
        """
        with aiohttp.ClientSession() as session:
            data = _ensure_bytes(data)
            data = b'<?xml version="1.0" encoding="utf-8" ?>\n' + data

            auth = '%s:%s' % (self.creds[0], self.creds[1])
            auth64 = _ensure_unicode(base64.b64encode(_ensure_bytes(auth))).replace('\n', '')
            headers['Authorization'] = 'Basic %s' % auth64

            headers['Content-type'] = 'application/xml; charset="utf-8"'
            headers['Content-length'] = str(len(data))
            with aiohttp.Timeout(timeout=timeout):
                async with session.post(self.url + '/cimom', data=data, headers=headers) as resp:
                    content = await resp.read()
                    l.debug('Reply from {}: Status {}, content {}'.format(self.url, resp.status, content))
        return content
```
and overriding `imethodcall`, substituting the below lines (590-591 in cim_operations.py):
```python
-       reply_xml = wbem_request(
-           self.url, req_xml.toxml(), self.creds, headers,

+       reply_xml = await self.wbem_request(
+            req_xml.toxml(), headers,
```

I am using `aiohttp` rather than `requests`, and I had to make several other changes to use the `async/await` syntax but you get the idea.  I also converted headers to a dictionary rather than a list of strings.  However, only the above changes would be really necessary to drop in `requests` to replace `cim_http.wbem_request`.

I'm not saying that's the direction that the package should go in, but for users I think it's easier than maintaining a fork.

**Comment by jealous** [comment]
Thanks for the sharing, James!

**Comment by KSchopmeyer** [comment]
Karl will work through the various issues with this move and then we can make a decision.

**Comment by KSchopmeyer** [comment]
We agreed to move this issue out to the next release.

**Comment by andy-maier** [comment]
I am starting with migrating to using the requests module.

**Comment by andy-maier** [comment]
PR #2023 migrates to use of the requests package, for both Python 2 and Python 3. This PR will get into master (i.e. the upcoming 1.0.0 version), and will not be rolled back. This change will remove functionality such as local logon or UNIX socket based local connectivity.
